### ISSUE TITLE
[💡 FEATURE REQUEST]: Configuration to UNIX socket permissions

### ISSUE THREAD

**Issue by Starfox64** [roadrunner-server/roadrunner#1789]
### Plugin

None

### I have an idea!

Add a configuration option to set the file permissions and ownership of UNIX sockets created for RPC, FastCGI, etc.

Right now it simply uses de default umask of the system which makes the socket inaccessible to non owners.

**Comment by rustatian** [comment]
Hey @Starfox64 👋 
Yeah, currently, RR uses the `parent` (RoadRunner's) process permissions to create sockets.
RPC... hm, are you working with RPC/FCGI via UNIX sockets?

**Comment by Starfox64** [comment]
Yes, I've switched to unix sockets for RPC and FastCGI, the performance increase is quite significant.

In my setup I have NGINX talking to RR's FastCGI and on the PHP side I use the KV plugin as a Laravel cache store:
- **TCP**: 220 rqps
- **UDS**: 306 rqps

*Side note*: I was quite surprised by this, as before I was using a Redis Sentinel cluster over the network (may have been on the same node, I didn't get to verify) was even faster (~320 rqps). I think there might be some improvement oportunities in the KV implementation.

**Comment by rustatian** [comment]
Ah, got u @Starfox64 👍 
Yeah, `UNIX` sockets are faster than `TCP`, and `fastcgi` handler is generally faster than a `proxy_pass`. I'll think about precise configuration of the permissions. Most likely, I'll spread the `server.uid`/`server.gid` options to all external file descriptors created by RR.

**Comment by karabanov** [comment]
Hi, I have the same issue
I hope it will resolve in near time

Only this needs to be done as in PHP-FPM, so that the rights to the socket can be assigned separately, since there is no need for a third-party user such as www-data or nginx to give write rights to all files it is enough for them to give write rights only to the unix socket and only read rights to all other files.

**Comment by Crone1331** [comment]
Yes, this admin says everything correctly and this is how it should be done.

**Comment by mikevmk** [comment]
> rights to the socket can be assigned separately

second to that

**Comment by IvanRazbegaev** [comment]
I agree with all that @karabanov wrote, this devops is right

**Comment by rustatian** [comment]
Why do you need to start RR from the root and have separate permission for the socket? Why not to start RR from the `www-something` user and automatically have inherited permissions for the socket?

**Comment by karabanov** [comment]
@rustatian not from `root`, from a regular user and do not assign write permissions to files to anyone except this user

In general, it would be very good if it were possible to assign rights to the unix socket separately so that everything would look like this:

- RoadRunner is launched on behalf of the owner `user:user` and the files would be assigned the following permissions `-rw-r--r--`
- Unix socket was created on behalf of the owner `user:www-data` and it was assigned the rights `srw-rw----`

I will try to explain this necessity

For example, now I am trying to run RoadRunner and make it listen to the socket. To do this, I need to run it on behalf of a regular user who is also a member of the **www-data** group, and also use an entrypoint script to set the correct `umask` and assign such permissions to the created Unix socket so that nginx can write to it, because if this is not done, the `srwxr-xr-x` permissions will be set on the socket, but it should be `srw-rw----`

Of course, my actions will make everything work, but with such a configuration, the user www-data will be able to write not only to the socket but also to other files, which I would not want even hypothetically

In PHP-FPM, the rights to the Unix socket are assigned separately (although it is done a little differently there) and nginx can only write to the socket, and all other files are read-only.

**Comment by rustatian** [comment]
Thanks for the detail explanation 👍 
Yeah, I agree that having precise permissions for the socket would be beneficial. Everything related to the socket creation in RR laying in a separate package [link](https://github.com/roadrunner-server/tcplisten), so, won't be difficult to implement this.

**Comment by karabanov** [comment]
Hi

This comment is just to draw attention to this issue - we would like to be able to assign rights to a unix-socket ourselves, as discussed in previous comments

**Comment by rustatian** [comment]
Hey @karabanov 👋🏻 
Yeah, I will try to allocate time for this feature in the nearest future. Sorry for the inconvenience, but I'm working on this project in my free time, so, delays are expected 😃

**Comment by rustatian** [comment]
@karabanov ,
Could you please share a PHP-FPM sample configuration (with permissions) you wanted to approximately resemble in RR (since I'm far away from PHP)?
Because I think this case might be solved a little bit differently for the fcgi (HTTP plugin).

**Comment by karabanov** [comment]
> [@karabanov](https://github.com/karabanov) , Could you please share a PHP-FPM sample configuration (with permissions) you wanted to approximately resemble in RR (since I'm far away from PHP)? Because I think this case might be solved a little bit differently for the fcgi (HTTP plugin).

Here is an example of such a config

```
[app]
; Standard
user = app
group = app
listen = /var/run/php8.3-fpm.app.sock
listen.owner = www-data
listen.group = www-data
listen.mode = 0600

... skipped ...
```

The goal is to grant read/write permissions only to the Unix socket - other files will not be writable by the user **www-data**

**Comment by rustatian** [comment]
Fixed, now it is possible to set mode (via umask, like 0777) + uid + gid for every Unix socket in RR.
